### PR TITLE
Storage: bind storage_type to each file that is saved

### PR DESCRIPTION
**Pull request by monkeyiq**
The addition of the new chunked storage type creates some issues when folks might want to migrate to using it while still serving old files that have already been transferred.

This PR will require a database upgrade to work;
```
php /opt/filesender/scripts/upgrade/database.php
```

Note that I have set the default storage for existing transfers to 'Filesystem'. If you have been using the chunked storage type then you will need to update the database with the below command to override the default setting, note use files as the table in postgresql and Files in MySQL.
```
update files set storage_type = 2 ;
```

Since each file saves the int value for the storage_type that was active when it was created you can more easily change the storage_type in the future while still serving existing files.

### CODE REVIEW

**Comment by monkeyiq** [comment]
This relates to https://github.com/filesender/filesender/issues/136

**Comment by jornane** [comment]
Interesting solution enumerating all possible storage classes. I would think that this might cause issues with parties implementing their own storage, because they would have to add a non-overlapping number. Have you designed an addressing plan for storage types?

Have you considered simply storing the class name of the storage type in the database as a varchar?

**Comment by madsi1m** [comment]
+1 in class name

On 10 Oct 2017, at 19:01, Jørn Åne <notifications@github.com<mailto:notifications@github.com>> wrote:


Interesting solution enumerating all possible storage classes. I would think that this might cause issues with parties implementing their own storage, because they would have to add a non-overlapping number. Have you designed an addressing plan for storage types?

Have you considered simply storing the class name of the storage type in the database as a varchar?

**Comment by monkeyiq** [comment]
TLDR; I personally wouldn't choose to varchar it. But since folks want that I will move to it.

My addressing plan for the future was to use a lookup table in the database
```
storage_types
-----------+----
id         | int primary key autoincr
class_name | varchar(100)
```

The Files.storage_type would refer to a storage_types.id. The actual value of that ID is not important. No addressing plan required, nobody has to care about getting a number. A script run once per code update or on request with third party storage classes would ensure that every storage class has an entry in the lookup table.

then a view on Files could integrate the id->class_name resolution and return that as filesView.storage_class_name which would be the varchar. The important part here is you are only actually storing about 4 bytes instead of 120+ per tuple. I only used the php lookup table for now because there is no view support in the db layers and we only recently went from one to two storage_types in total.

Using an int or 4 byte int with a small lookup table (fully RAM cached) is so much more efficient than storing a varchar(30) per tuple. In 4 byte unicode that becomes 120 bytes+ of storage per tuple. While I agree that disk space is cheap it will also lead to larger tables and less efficient indexes and caches.

**Comment by jornane** [comment]
Lookup in an on-the-fly populated database table is a good solution, lookup using hard-coded ENUMs is problematic I think.

 By using ENUMs as a "temporary" solution, you'd have to write a migrate script later to translate these ENUMs to database entries. That would make future setups a bit hairy, because then there would be some "special" numbers that must map to known storage types.

My opinion:

ENUM < VARCHAR < Lookup-DB

**Comment by monkeyiq** [comment]
I feel very confident that I could arrange to insert two tuples into a newly created table during migration. 

The database table would never have been populated "on the fly", databases are kind of good at storing information, so I would trust it to store these two tuples for the join table.

**Comment by madsi1m** [comment]
My feeling is when adding an extra custom storage type not already included in the code base and it getting an id because auto increment, followed by code being updated by adding another storage type, the ids would not be the same on everyone’s deployments, which isn’t bad, just slightly harder to support when other installs get stuck. I think sticking to the storage class name is simple enough and not taxing on the db. Also I don’t want to have to insert into a storage table each time I write a new storage backend.

On 22 Oct 2017, at 12:28, monkeyiq <notifications@github.com<mailto:notifications@github.com>> wrote:


I feel very confident that I could arrange to insert two tuples into a newly created table during migration.

The database table would never have been populated "on the fly", databases are kind of good at storing information, so I would trust it to store these two tuples for the join table.

**Comment by monkeyiq** [comment]
there are easier pickings to db updates to improve performance.
